docs/rfc/rfc-026-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,436 @@
+
+- What if nodes run different versions of Tendermint and banning is not supported in one of the versions? 
+
+- Reserving response codes can be problematic for existing applications that may have reserved these codes for internal purposes withtou being aware that this causes a ban now. 
+ 
+
+## Implementation considerations
+
+**Indicating a new type of `CheckTx` failure**
+
+The initial proposal is to reserve a special response code to indicate that the transaction could never have been valid. 
+Due to concerns of this being a breaking change for applications that have already reserved this code for internal
+purposes, there is an alternative implementation: expanding `ResponseCheckTx` with an additional field. 
+This field `neverValidTx` would be `false` by default. If a transaction could never have been valid,
+in addition to indicating this with a non-zero response code from `CheckTx`, the application would set this field value. 
+
+**Adding support for peer banning**
+
+When a transaction fails `CheckTx`, it is not stored in the mempool but **can** be stored in the cache. If it is in the cache, it cannot be resubmitted again (as it will be discovered in the cache and not checked again). These two scenarios require a different implementation of banning in case `CheckTx` failed. 
+
+In both cases we need to keep track of the peers that sent invalid transactions. If invalid transactions are cached, 
+we also need to keep track of the `CheckTx` response code for each transaction. Currently the `ResponseCheckTx` code is checked in `resCbFirstTime` of the mempool. 
+If invalid transactions are kept in the cache, the check is ran only when a transaction is 
+seen for the first time. Afterwards, the  transaction is cached, to avoid running `CheckTx` on transactions already checked. 
+Thus when a transaction is received from a peer, if it is in the cache,
+`CheckTx` is not ran again, but the peers' ID is addded to the list of peers who sent this particular transaction.
+These transactions are rechecked once a block is committed to verify that they are still valid. 
+
+If invalid transactions are not kept in the cache, they can be resubmitted multiple times, and `CheckTx` will be executed on them upon submission. 
+Therefore we do not need to remember the previous response codes for these transactions.
+
+In summary, if we want to support banning peers based on the frequency with which they submit invalid transactions, we need to have **additional datastructures**:
+1. One to keep track of past invalid transactions
+2. A datastructure to differentiate between valid and invalid *cached* transactions. If the `KeepInvalidTxsInCache` configuration parameter is not set, this datastructure
+   is not needed. 
+
+We propose two ways to implement peer banning based on the result of `CheckTx`:
+
+1. Introduce banning when transactions are received
+2. Adapt the recheck logic to support this
```

Comment:
The problem of 2. is that we don't know who to punish/ban in the case the transaction has been received from multiple peers.